sdk/src/test/scala/com/horizen/account/wallet/AccountWalletTest.scala

```diff
@@ -195,4 +191,79 @@ class AccountWalletTest
+    Mockito.when(mockedSecretStorage.add(ArgumentMatchers.any[Secret])).thenReturn(Success(mockedSecretStorage))
+    Mockito.when(mockedSecretStorage.storeNonce(ArgumentMatchers.anyInt(), ArgumentMatchers.any[Array[Byte]])).thenReturn(Success(mockedSecretStorage))
+
+    // Prepare block ID and corresponding version
+    val blockId = new Array[Byte](32)
+    Random.nextBytes(blockId)
+
+    Mockito.when(mockedSecretStorage.getNonce(ArgumentMatchers.any())).thenReturn(Some(2))
+    val privateKey25519Creator = PrivateKey25519Creator.getInstance()
+
+    val result3 = accountWallet.generateNextSecret(privateKey25519Creator)
+    assertTrue("Generation of first key should be successful", result3.isSuccess)
+    val secret3 = result3.get._2
+    storageList += secret3
+    storageList -= secret1
+
+    Mockito.when(mockedSecretStorage.getAll).thenReturn(storageList.toList)
+    val result4 = accountWallet.generateNextSecret(privateKey25519Creator)
+
+    assertTrue("Generation of second key should be successful", result4.isSuccess)
+    val secret4 = result4.get._2
+
+    assertEquals("keys should not be the same", secret3, secret4)
+  }
+
+  @Test
+  def testGenerateSecretsOfDifferentDomainsIndependent(): Unit = {
+    val mockedSecretStorage: SidechainSecretStorage = mock[SidechainSecretStorage]
+
+    val storageList = ListBuffer[Secret]()
+    val secret1 = getPrivateKey25519("seed1".getBytes())
+    val secret2 = getSchnorrKey("seed2".getBytes())
+    storageList += secret1
+    storageList += secret2
+
+    val privateKey25519Creator = PrivateKey25519Creator.getInstance()
+    val schnorrKeyCreator = SchnorrKeyGenerator.getInstance()
+    Mockito.when(mockedSecretStorage.getNonce(privateKey25519Creator.salt())).thenReturn(Some(1))
+    Mockito.when(mockedSecretStorage.getNonce(schnorrKeyCreator.salt())).thenReturn(Some(0))
+
+    Mockito.when(mockedSecretStorage.getAll).thenReturn(storageList.toList)
+    Mockito.when(mockedSecretStorage.add(ArgumentMatchers.any[Secret])).thenReturn(Success(mockedSecretStorage))
+    Mockito.when(mockedSecretStorage.storeNonce(ArgumentMatchers.anyInt(), ArgumentMatchers.any[Array[Byte]])).thenReturn(Success(mockedSecretStorage))
+
+    val accountWallet = new AccountWallet(
+      "seed".getBytes(),
+      mockedSecretStorage)
+    val result3 = accountWallet.generateNextSecret(privateKey25519Creator)
+    assertTrue("Generation of first key should be successful", result3.isSuccess)
+    val secret3 = result3.get._2
+    storageList += secret3
+    storageList -= secret2
+
+    Mockito.when(mockedSecretStorage.getAll).thenReturn(storageList.toList)
+    val result4 = accountWallet.generateNextSecret(schnorrKeyCreator)
+
+    assertTrue("Generation of second key of different type should be successful", result4.isSuccess)
+    val secret4 = result4.get._2
+
+    assertNotEquals("The deleted Schnorr key and the generated Schnorr key after deleted of that one should not be the same", secret2, secret4)
```

Comment:
maybe I miss something, but why do we expect secret2 and secret4 to be different?
Note: at the moment `schnorrKeyCreator` is not deterministic, so better to test with `privateKey25519Creator`.

Reply:
Changed test logic

Reply:
1. Storage {Key25519_1, SchnorrKey_1, Key25519_2, Key25519_3,  SchnorrKey_2}
25519Nonce = 2
SchnorrNonce = 1

2. We add SchnorrKey_3
assert Mockito.when(storeNonce) => thenAnswer(assert(Nonce==2 && KeyTypeSalt=="Schnorr"))

3. We add Key25519_4
assert Mockito.when(storeNonce) => thenAnswer(assert(Nonce==3 && KeyTypeSalt=="25519"))

4. We add SchnorrKey_4
assert Mockito.when(storeNonce) => thenAnswer(assert(Nonce==3 && KeyTypeSalt=="Schnorr"))

5. We add Key25519_5
assert Mockito.when(storeNonce) => thenAnswer(assert(Nonce==4 && KeyTypeSalt=="25519"))